java/core/src/main/java/co/worklytics/psoxy/gateway/impl/InboundWebhookHandler.java

```diff
@@ -0,0 +1,197 @@
+                    .build();
+            }
+        } else {
+            try {
+                authToken = Optional.of(SignedJWT.parse(authorizationHeader.get()));
+            } catch (ParseException e) {
+                return HttpEventResponse.builder()
+                    .statusCode(HttpStatus.SC_BAD_REQUEST)
+                    .body("Authorization header is present, but could not be parsed as a JWT: " + e.getMessage())
+                    .build();
+            }
+
+            // verify authorization header is signed with one of the acceptable public keys
+            Optional<String> validationError = this.validate(authToken.get(), acceptableAuthKeys());
+
+            if (validationError.isPresent()) {
+                return HttpEventResponse.builder()
+                    .statusCode(HttpStatus.SC_UNAUTHORIZED)
+                    .body("Authorization header is present, but validation failed: " + validationError.get())
+                    .build();
+            }
+
+            Map<String, Object> claims;
+            try {
+                claims = authToken.get().getJWTClaimsSet().getClaims();
+            } catch (ParseException e) {
+                claims = null;
+            }
+
+            if (!webhookSanitizer.verifyClaims(request, claims)) {
+                return HttpEventResponse.builder()
+                    .statusCode(HttpStatus.SC_UNAUTHORIZED)
+                    .body("Claims in Authorization header do not match request")
+                    .build();
+            }
+        }
+
+        // what if Authorization header sent, but not intended for Psoxy?
+        // AWS API Gateway / Lambda AWS IAM authorizer will STRIPE the Authorization header, not pass it to the Lambda function - so OK
+        // we'll get a plain 'Authorization' header ONLY in case where Auth=NONE on the Lambda Function URL or the API Gateway
+
+        if (!webhookSanitizer.canAccept(request)) {
+            return HttpEventResponse.builder()
+                .statusCode(HttpStatus.SC_BAD_REQUEST)
+                .build();
+        }
+
+
+
+        ProcessedContent sanitized = webhookSanitizer.sanitize(request);
+
+        output.write(sanitized);
+
+        return HttpEventResponse.builder()
+            .statusCode(HttpStatus.SC_OK)
+            .build();
+    }
+
+    /**
+     * returns a failure message if invalid, or empty otherwise
+     * @param signedJWT
+     * @param publicKeys
+     * @return optional with the failure, if any
+     */
+    @SneakyThrows
+    public Optional<String> validate(SignedJWT signedJWT, Collection<RSAPublicKey> publicKeys) {
+
+        boolean signatureValid = publicKeys.stream()
+            .filter(k -> {
+                JWSVerifier verifier = new RSASSAVerifier(k);
+                try {
+                    return signedJWT.verify(verifier);
+                } catch (JOSEException e) {
+                    throw new RuntimeException(e);
```

Comment:
you may have 2+ keys, right? while rotating, I wouldn't err here, just in case the first one doesn't work, should try with next one

Reply:
may have multiple, depends on approach. I thought returns `false` when signature can't be verified, but yeah - I agree that docs unclear about when `JOSEException` is thrown, so we can treat as equivalent to verification returning `false`.